enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj

```diff
@@ -0,0 +1,164 @@
+             (if (= capacity (count tree))
+               (recur (disj tree (first tree)) candidate)
+               (conj tree [gated_at id])))]
+    {:capacity capacity
+     :tree     (reduce rf tree update-candidates)}))
+
+(defn indexing-step [pgvector index-metadata index indexing-state & {:keys [ignore-filter]}]
+  (let [{:keys [watermark candidate-filter]} @indexing-state
+
+        move-to-next-watermark
+        (fn [poll-result]
+          (let [next-watermark (semantic.gate/next-watermark watermark poll-result)]
+            (vswap! indexing-state assoc :watermark next-watermark)
+            (semantic.gate/flush-watermark! pgvector index-metadata index next-watermark)))
+
+        {:keys [update-candidates] :as poll-result}
+        (semantic.gate/poll pgvector index-metadata watermark
+                            :limit poll-limit
+                            :lag-tolerance lag-tolerance)]
+    (if-some [novel-candidates (if ignore-filter
+                                 (seq update-candidates)
+                                 (seq (remove-redundant-candidates candidate-filter update-candidates)))]
+      (let [documents-query   {:select [:model :model_id :document]
+                               :from   [(keyword (:gate-table-name index-metadata))]
+                               :where  [:in :id (sort (map :id novel-candidates))]}
+            documents-sql     (sql/format documents-query :quoted true)
+            gate-docs         (jdbc/execute! pgvector documents-sql {:builder-fn jdbc.rs/as-unqualified-lower-maps})
+            updates           (filter :document gate-docs)
+            deletes           (remove :document gate-docs)
+            updated-documents (map semantic.gate/gate-doc->search-doc updates)]
+        (log/infof "Found gate updates: %d updates, %d deletes" (count updates) (count deletes))
+        (when (seq updated-documents)
+          (semantic.index/upsert-index! pgvector index updated-documents))
+        (doseq [[model deletes] (group-by :model deletes)]
+          (semantic.index/delete-from-index! pgvector index model (map :model_id deletes)))
+        (vswap! indexing-state assoc
+                :last-seen-change (Instant/now)
+                :last-indexed-count (count novel-candidates)
+                :last-poll-count (count update-candidates)
+                :candidate-filter (update-candidate-filter candidate-filter novel-candidates))
+        (move-to-next-watermark poll-result))
+      (do
+        (vswap! indexing-state assoc
+                :last-indexed-count 0
+                :last-poll-count (count update-candidates))
+        (move-to-next-watermark poll-result)))))
+
+;; having a var is handy for tests
+(defn- sleep [ms]
+  (Thread/sleep (long ms)))
+
+(defn on-indexing-idle [indexing-state]
+  (let [{:keys [last-poll-count last-indexed-count]} @indexing-state
+        novelty-ratio (if (zero? last-poll-count) 0 (/ last-indexed-count last-poll-count))]
+    (cond
```

Comment:
this survived too long from sketch stage but its too cute I'll probably tight-loop until buffer is full, poll cost ought to be marginal compared to embed/index and we can revisit if `:limit` ends up being huge for some reason